sycl/doc/extensions/proposed/sycl_ext_oneapi_barrier.asciidoc

```diff
@@ -0,0 +1,251 @@
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+:dpcpp: pass:[DPC++]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+[%hardbreaks]
+Copyright (C) 2023-2023 Intel Corporation.  All rights reserved.
+
+Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 7 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension defines a `sycl::ext::oneapi::experimental::barrier` class
+aligned with the `std::barrier` class defined in {cpp}20. This extension may
+also be considered a generalization of the existing
+link:../experimental/sycl_ext_oneapi_cuda_async_barrier.asciidoc[sycl_ext_oneapi_cuda_async_barrier]
+extension for non-CUDA backends.
+
+`sycl::ext::oneapi::experimental::barrier` is _not_ a general replacement for
+`sycl::group_barrier`, and is intended for usage by experts seeking to
+synchronize subsets of threads of execution that do not necessarily correspond
+to groups of work-items.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_BARRIER` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== `barrier` class
+
+The `sycl::ext::oneapi::experimental::barrier` class has the same interface as
+the `std::barrier` class from {cpp}20, with an additional `Scope` template
+parameter denoting a memory scope broad enough to contain all threads of
+execution that are able to use the barrier.
+
```

Comment:
Does the memory scope here determine the storage of the barrier that makes it shared/accessible by the threads that can access that memory?
For instance, if the memory scope is WI, no other thread can use that barrier, right?

Reply:
The storage for the barrier is determined by the developer when they allocate the barrier object -- a barrier used to synchronize work-items in different work-groups would have to be allocated in global memory (e.g. with `sycl::malloc_device`, or a global variable) whereas a barrier used to synchronize work-items in the same work-group _could_ be allocated in group-local memory (e.g. with `syclex::group_local_memory`).

The scope is orthogonal, and behaves similarly to the `memory_scope` argument used by `sycl::atomic_ref`.  For example, if the implementation knows that all of the work-items it is synchronizing are in the same work-group, it can generate narrower fence.  Such a work-group barrier could be stored in global memory or local memory.

I think that a barrier with a memory scope set to `memory_scope::work_item` could only be used by a single work-item, but I'm not 100% sure whether it would necessarily have to be the same work-item every time...  It wouldn't be a very useful thing to have, in any case.

I've attempted to clarify this in 3b8e80e3e8cc5077043c0b3705af86bf2de8b3ba.